core/src/layers/observe/metrics.rs

```diff
@@ -132,18 +157,60 @@ impl<A: Access, I: MetricsIntercept> Layer<A> for MetricsLayer<I> {
     fn layer(&self, inner: A) -> Self::LayeredAccess {
         let info = inner.info();
 
+        // Update http client with metrics http fetcher.
+        info.update_http_client(|client| {
+            HttpClient::with(MetricsHttpFetcher {
+                inner: client.into_inner(),
+                info: info.clone(),
+                interceptor: self.interceptor.clone(),
+            })
+        });
+
         MetricsAccessor {
-            inner: Arc::new(inner),
+            inner,
             info,
             interceptor: self.interceptor.clone(),
         }
     }
 }
 
+/// The metrics http fetcher for opendal.
+pub struct MetricsHttpFetcher<I: MetricsIntercept> {
+    inner: HttpFetcher,
+    info: Arc<AccessorInfo>,
+    interceptor: I,
+}
+
+impl<I: MetricsIntercept> HttpFetch for MetricsHttpFetcher<I> {
+    async fn fetch(&self, req: http::Request<Buffer>) -> Result<http::Response<HttpBody>> {
+        // Extract context from the http request.
+        let size = req.body().len();
+        let op = req.extensions().get::<Operation>().copied();
```

Comment:
one question: where to insert the operation into the request extensions?

Reply:
Covered by https://github.com/apache/opendal/issues/5774